src/main/kotlin/io/quarkus/code/service/PlatformService.kt

```diff
@@ -0,0 +1,105 @@
+package io.quarkus.code.service
+
+import io.quarkus.code.misc.QuarkusExtensionUtils.processExtensions
+import io.quarkus.code.config.ExtensionProcessorConfig
+import io.quarkus.devtools.project.QuarkusProjectHelper
+import io.quarkus.code.model.CodeQuarkusExtension
+import java.util.HashMap
+import io.quarkus.registry.catalog.PlatformCatalog
+import java.time.LocalDateTime
+import javax.enterprise.event.Observes
+import io.quarkus.runtime.StartupEvent
+import io.quarkus.scheduler.Scheduled
+import java.time.ZoneOffset
+import java.time.ZoneId
+import io.quarkus.registry.RegistryResolutionException
+import kotlin.Throws
+import java.util.logging.Logger
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class PlatformService {
+    @Inject
+    lateinit var extensionProcessorConfig: ExtensionProcessorConfig
+    private val catalogResolver = QuarkusProjectHelper.getCatalogResolver()
+    var streamCatalogMap: MutableMap<String, List<CodeQuarkusExtension>> = HashMap()
+    var platformCatalog: PlatformCatalog? = null
+    var lastUpdated: LocalDateTime? = null
+
+    fun onStart(@Observes e: StartupEvent?) {
+        reloadCatalogs()
+    }
+
+    @Scheduled(cron = "{reload.catalogs.cron.expr}")
+    fun reloadCatalogs() {
+        try {
+            platformCatalog = catalogResolver.resolvePlatformCatalog()
+            populateExtensionCatalogMaps()
+            lastUpdated = LocalDateTime.now(ZoneOffset.UTC as ZoneId)
+        } catch (e: RegistryResolutionException) {
+            LOG.warning("Could not reload catalogs [" + e.message + "]")
+        }
+    }
+
+    val extensionCatalog: List<CodeQuarkusExtension>?
+        get() {
+            val defaultPlatformKey = platformCatalog!!.recommendedPlatform.platformKey
```

Comment:
Wouldn't using a local copy of the platformCatalog be safer, e.g.
````
val pc = platformCatalog;
val defaultPlatformKey = pc!!.recommendedPlatform.platformKey
...
````
in case the platformCatalog gets updated in middle?

Reply:
Yes we should be really careful about the state and concurrency now that it gets refreshed, also maybe it would be safer to use `AtomicReference` for mutable state fields?

Reply:
This is done (not yet committed)